service/topology_coordinator.hh

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include <memory>
+#include <chrono>
+#include <cstdint>
+
+#include <seastar/core/future.hh>
+#include <seastar/core/sharded.hh>
+
+#include "raft/raft.hh"
+#include "service/topology_state_machine.hh"
+
+namespace db {
+class system_keyspace;
+class system_distributed_keyspace;
+}
+
+namespace gms {
+class gossiper;
+}
+
+namespace netw {
+class messaging_service;
+}
+
+namespace locator {
+class shared_token_metadata;
+}
+
+namespace replica {
+class database;
+}
+
+namespace raft {
+class server;
+}
+
+namespace service {
+
+class raft_group0;
+class tablet_allocator;
+
+class topology_coordinator {
+public:
+    virtual ~topology_coordinator() {}
```

Comment:
Why make it a virtual interface? It suggests that there might be multiple implementations. But there is and always will be only one.

WDYT about using pimpl pattern instead? Like with `group0_guard`.
```cpp
class group0_guard {
    struct impl;
    std::unique_ptr<impl> _impl;

    group0_guard(std::unique_ptr<impl>);
public:
    ~group0_guard();
```

Reply:
Even this is not needed IMO. Since topology_coordinator exists as a unique_ptr only I think the only thing that is needed is to make destructor available in the header.

Reply:
> WDYT about using pimpl pattern instead? Like with `group0_guard`.

I usually don't like the pimpl idiom because of the boilerplate that is needed for each exposed method, but considering that `topology_coordinator` only needs to expose `::run()` then it shouldn't be too bad.